### PR TITLE
Add Extra Features to WCF Extension

### PR DESCRIPTION
**Pull request by SimonaLiao**
This PR adds extra features & resolve some of the code review comments based on the previous WCF PR.
Here is the list of added features:
1. Supports WCF project with multiple services
2. Supports `ServiceCredentials`, `ServiceDebug`, `ServiceMetadata`... configuration properties
3. Support adding Windows authentication services to DI container

For code review comments, this PR added two new classes `FilePath` and `ConfigContext` to have strongly typed variables.
Addresses #1270

### CODE REVIEW

**Review by brandonh-msft — CHANGES_REQUESTED** [review]
👍🏻 
Blocking until 1+ WCF team members approve

File: src/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater/ConfigContext.cs [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+using System.Xml.Linq;
+
+namespace Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater
+{
+    public class ConfigContext
+    {
+        public Dictionary<string, ServiceSpecificContext> ServiceContext { get; }
+
+        public bool NetTcpCertificate { get; }
+
```
Comment (mconnew) [inline src/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater/ConfigContext.cs:16]: Try to name bool properties like they are answering a question. For example you already have HasMetadata. I presume this is a bool to specify if NetTcp is using certificates, so maybe NetTcpUsesCertificates. Same with WindowsAuthentication

File: src/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater/ConfigContext.cs [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+        public class ServiceSpecificContext
+        {
+            public Dictionary<string, Uri> SchemeToAddressMapping { get; }
+
+            public MetadataType MetadataType { get; }
+
+            public Dictionary<string, string> ServiceDebug { get; }
+
+            public HashSet<string> Bindings { get; }
+
```
Comment (mconnew) [inline src/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater/ConfigContext.cs:30]: Is this the set of binding types, or binding configuration names? Can you make the name a little more description?

File: src/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater/ConfigContext.cs [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+            public bool ServiceCertificate { get; }
+
```
Comment (mconnew) [inline src/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater/ConfigContext.cs:32]: Same thing about the naming as above, is this UsesServiceCertifate? Or ServiceCertificateConfigured?

File: src/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater/ConfigContext.cs [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+            public Dictionary<string, string> ServiceCredentials { get; }
+
```
Comment (mconnew) [inline src/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater/ConfigContext.cs:34]: Is this holding the set of properties for the service credentials? If so, maybe call it ServiceCredentialProperties. If it's something else, can you try and find a more descriptive name.

File: src/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater/ConfigUpdater.cs [hunk 5]
```diff
@@ -28,6 +29,7 @@ public ConfigUpdater(XDocument doc, ILogger<ConfigUpdater> logger)
         {
             _config = doc;
             _logger = logger;
+            SetServiceAndBehaviorNames(_logger);
         }
```
Comment (mconnew) [inline src/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater/ConfigUpdater.cs:33]: You don't need to pass the ILogger, you've saved it in a field and the method is an instance method so can access it.

File: src/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater/SourceCodeUpdater.cs [hunk 6]
```diff
@@ -50,6 +50,7 @@ public SyntaxNode UpdateDirectives()
             }
 
             var root = _programFile.GetRoot();
+            var templateRoot = CSharpSyntaxTree.ParseText(Constants.TemplateUsing).GetRoot();
```
Comment (mconnew) [inline src/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater/SourceCodeUpdater.cs:54]: It doesn't look like templateRoot is used in this method.

File: tests/tool/Integration.Tests/IntegrationScenarios/WCFSample/Upgraded/wcf.config [hunk 7]
```diff
@@ -2,7 +2,7 @@
 <configuration>
   <system.serviceModel>
     <services>
-      <service name="WcfServiceLibrary1.Service1">
+      <service name="WcfServiceLibrary1.Service1" behaviorConfiguration="">
         <!--The host element is not supported in configuration in CoreWCF. The port that endpoints listen on is instead configured in the source code.-->
```
Comment (mconnew) [inline tests/tool/Integration.Tests/IntegrationScenarios/WCFSample/Upgraded/wcf.config:6]: Does everything still work when behaviorConfiguration isn't specified as well as the behavior name attribute? The usual case will be them not being specified in the config file.

File: src/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater/ConfigContext.cs (hunk 2 above)
Comment (SimonaLiao) [reply src/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater/ConfigContext.cs:30]: This is the set of binding types, and yes I will improve the name!

File: tests/tool/Integration.Tests/IntegrationScenarios/WCFSample/Upgraded/wcf.config (hunk 7 above)
Comment (SimonaLiao) [reply tests/tool/Integration.Tests/IntegrationScenarios/WCFSample/Upgraded/wcf.config:6]: When the behavior name is not specified, I will set them to be empty string so they can still be passed into different methods. It's done in the SetServiceAndBehaviorNames() method.

File: src/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater/WCFUpdateStep.cs [hunk 8]
```diff
@@ -242,25 +242,42 @@ private void FindPath(IProject project)
...
+                // only updates the namespace for .cs files does not include clients
+                var usingDirectivesUpdates = new List<string>();
                 foreach (var directive in directives)
                 {
-                    Logger.LogDebug($"This .cs file: {directive} needs using directives updates. Adding the path to collection.");
+                    var root = CSharpSyntaxTree.ParseText(File.ReadAllText(directive)).GetRoot();
+                    if (ContainsIdentifier(root, "ChannelFactory") || ContainsIdentifier(root, "ClientBase"))
+                    {
+                        Logger.LogDebug($"This .cs file: {directive} needs using directives updates. Adding the path to collection.");
+                    }
```
Comment (mconnew) [inline src/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater/WCFUpdateStep.cs:260]: I don't think this log message is correct here as it's the same message as the else case.

File: tests/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater.Test/SourceCodeUpdaterTest.cs [hunk 9]
```diff
@@ -3,6 +3,7 @@
 
 using Microsoft.CodeAnalysis.CSharp;
 using Microsoft.Extensions.Logging.Abstractions;
+using Newtonsoft.Json.Bson;
 using Xunit;
```
Comment (mconnew) [inline tests/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater.Test/SourceCodeUpdaterTest.cs:7]: Was adding this using intentional?

File: tests/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater.Test/TestExpectedFiles/CredentialsHttpsTemplateCode.txt [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+                    int UA_placeHolder;
+                    host.Credentials.UserNameAuthentication.UserNamePasswordValidationMode = UserNamePasswordValidationMode.Custom;
+                    host.Credentials.UserNameAuthentication.CustomUserNamePasswordValidator = new MyType();
+                    host.Credentials.WindowsAuthentication.IncludeWindowsGroups = false;
+                });
+
+            });
+            
+            await app.StartAsync();
+            await app.StopAsync();
```
Comment (mconnew) [inline tests/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater.Test/TestExpectedFiles/CredentialsHttpsTemplateCode.txt:54]: This method is still void Main and not async Task Main. Either the await update is incorrect or the method signature is incorrect.

Comment (SimonaLiao) [reply tests/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater.Test/TestExpectedFiles/CredentialsHttpsTemplateCode.txt:54]: This and the following files are the updated version of template code, not the final expected source code from customer. So the void Main is just wrapping the template code so the SyntaxTree can parse it.

File: tests/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater.Test/SourceCodeUpdaterTest.cs (hunk 9 above)
Comment (SimonaLiao) [reply tests/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater.Test/SourceCodeUpdaterTest.cs:7]: Nope! I think probably it was auto added but I didn't really need this. Will remove it!

**Review by mconnew — APPROVED** [review]

**Comment by mconnew** [comment]
> 👍🏻 Blocking until 1+ WCF team members approve

@brandonh-msft, approved

**Review by brandonh-msft — APPROVED** [review]

**Pull request merged**